storage/postgreskv/client.go

```diff
@@ -332,58 +331,85 @@ func (client *Client) CompareAndSwapPath(ctx context.Context, bucket, key storag
 		return storage.ErrEmptyKey.New("")
 	}
 
-	tx, err := client.pgConn.BeginTx(ctx, nil)
-	if err != nil {
-		return Error.Wrap(err)
-	}
-
-	q := "SELECT metadata FROM pathdata WHERE bucket = $1::BYTEA AND fullpath = $2::BYTEA FOR UPDATE"
-	row := tx.QueryRow(q, []byte(bucket), []byte(key))
-	var val []byte
-	err = row.Scan(&val)
-	if err == sql.ErrNoRows {
-		// no row selected, so no need for transaction
-		txErr := Error.Wrap(tx.Commit())
-
-		if oldValue != nil {
-			return errs.Combine(storage.ErrKeyNotFound.New(key.String()), txErr)
+	if oldValue == nil && newValue == nil {
+		q := "SELECT metadata FROM pathdata WHERE bucket = $1::BYTEA AND fullpath = $2::BYTEA"
+		row := client.pgConn.QueryRow(q, []byte(bucket), []byte(key))
+		var val []byte
+		err = row.Scan(&val)
+		if err == sql.ErrNoRows {
+			return nil
 		}
-
-		if newValue == nil {
-			return txErr
+		if err != nil {
+			return Error.Wrap(err)
 		}
+		return storage.ErrValueChanged.New(key.String())
+	}
 
+	if oldValue == nil {
 		q := `INSERT INTO pathdata (bucket, fullpath, metadata) VALUES ($1::BYTEA, $2::BYTEA, $3::BYTEA)`
 		_, err = client.pgConn.Exec(q, []byte(bucket), []byte(key), []byte(newValue))
 		if err, ok := err.(*pq.Error); ok {
 			if err.Code == "23505" { // unique_violation
-				return errs.Combine(storage.ErrValueChanged.New(key.String()), txErr)
+				return storage.ErrValueChanged.New(key.String())
 			}
 		}
-		return txErr
-	}
-	if err != nil {
-		return Error.Wrap(errs.Combine(err, tx.Rollback()))
-	}
-
-	if !bytes.Equal(val, oldValue) {
-		return errs.Combine(storage.ErrValueChanged.New(key.String()), tx.Rollback())
+		return Error.Wrap(err)
 	}
 
 	if newValue == nil {
-		q := "DELETE FROM pathdata WHERE bucket = $1::BYTEA AND fullpath = $2::BYTEA"
-		_, err := tx.Exec(q, []byte(bucket), []byte(key))
+		q := `
+		WITH matching_key AS (
+			SELECT * FROM pathdata WHERE bucket = $1::BYTEA AND fullpath = $2::BYTEA
+		), updated AS (
+			DELETE FROM pathdata
+				USING matching_key mk
+				WHERE pathdata.metadata = $3::BYTEA
+					AND pathdata.bucket = mk.bucket
+					AND pathdata.fullpath = mk.fullpath
+				RETURNING 1
+		)
+		SELECT EXISTS(SELECT 1 FROM matching_key) AS key_present, EXISTS(SELECT 1 FROM updated) AS value_updated
+		`
```

Comment:
From this point on, the logic ought to be identical to the other branch, starting at line 402. The only difference between these cases should be the query string. (This case doesn't need any argument for `$4`, but having the extra argument passed to `QueryRow()` shouldn't hurt anything.) Would it make sense to combine these two cases?

Reply:
What you say makes sense, but `QueryRow()` actually fails with ` pq: got 4 parameters but the statement requires 3` due to the extra argument.

Reply:
Oh! I guess they do check that after all. Well, you could put a `$4` into the query in a way that doesn't have any actual effect (example: change the last line to have `EXISTS(SELECT $4 FROM matching_key)` but that would be confusing. Or maybe you could do the `QueryRow` calls in the two branches separately, but still combine the remainder of the logic?

Reply:
I went with including `QueryRow` in both branches. Chech the latest commit.